edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+
+    public class MetricsWorker : IDisposable
+    {
+        readonly IMetricsScraper scraper;
+        readonly IMetricsStorage storage;
+        readonly IMetricsUpload uploader;
+        readonly ISystemTime systemTime;
+        readonly AsyncLock scrapeUploadLock = new AsyncLock();
+        static readonly ILogger Log = Logger.Factory.CreateLogger<MetricsScraper>();
+
+        DateTime lastUploadTime = DateTime.MinValue;
+
+        // This acts as a local buffer. It stores the previous value of every metric.
+        // If the new value for that metric is unchanged, it doesn't write the duplicate value to disk.
+        Dictionary<int, Metric> metrics = new Dictionary<int, Metric>();
+
+        PeriodicTask scrape;
+        PeriodicTask upload;
+
+        public MetricsWorker(IMetricsScraper scraper, IMetricsStorage storage, IMetricsUpload uploader, ISystemTime systemTime = null)
+        {
+            this.scraper = scraper;
+            this.storage = storage;
+            this.uploader = uploader;
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void Start(TimeSpan scrapingInterval, TimeSpan uploadInterval)
+        {
+            this.scrape = new PeriodicTask(this.Scrape, scrapingInterval, scrapingInterval, Log, "Scrape");
+            this.upload = new PeriodicTask(this.Upload, uploadInterval, uploadInterval, Log, "Scrape");
+        }
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+
+                foreach (var moduleResult in await this.scraper.ScrapeAsync(cancellationToken))
+                {
+                    IEnumerable<Metric> scrapedMetrics = PrometheousMetricsParser.ParseMetrics(this.systemTime.UtcNow, moduleResult.Value);
```

Comment:
we should not hard-code Prometheus metric parser here; either moving this logic inside IScraper or create an interface for IMetricsParser.

Reply:
It makes sense to move it to the scraper